### PR TITLE
Revive introduction of libssh2_userauth_banner

### PR DESCRIPTION
**Pull request by palmin**
This PR tries to revive #274 by @dmiller-nmap and @bonsaiviking from late 2018.

It includes bounds checks as requested by @willco007 and merges cleanly with libssh 1.9.0.

### CODE REVIEW

**Comment by palmin** [comment]
Closing and reopening PR to trigger a Travis rebuild.

**Comment by willco007** [comment]
This looks OK to me, anyone else using this?

File: src/userauth.c [hunk 1]
```diff
@@ -189,6 +233,30 @@ libssh2_userauth_list(LIBSSH2_SESSION * session, const char *user,
...
+ * been made, this function returns NULL.
+ * libssh2_userauth_list makes a "none" authentication attempt and is
+ * sufficient to collect the pre-auth banner message.
+ *
+ * Banner ought to be UTF-8 encoded, and will be truncated to
+ * LIBSSH2_USERAUTH_MAX_BANNER bytes. Length will be returned in
+ * banner_len_out.
+ */
+LIBSSH2_API char *
+libssh2_userauth_banner(LIBSSH2_SESSION * session,
```
[the remark this answers is not in the document]
Comment (palmin) [reply src/userauth.c]: Should I change `libssh2_userauth_banner` to have a return code and output parameter for the banner?

[the remark this answers is not in the document]
Comment (willco007) [reply src/userauth.c]: That would be more consistent with the existing APIs.

**Comment by palmin** [comment]
Travis build is failing after compiling tests but while running them and I don't really know how to fix this.

It looks like it is some kind of [Docker problem](https://travis-ci.org/libssh2/libssh2/jobs/646874665#L573) and I'm not experiencing this when running tests locally. 

Any advice on things to attempt?

File: src/userauth.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (palmin) [reply src/userauth.c]: Does this look better?

https://github.com/libssh2/libssh2/pull/444/files#diff-a46a7c531a0ed0c1f0d7e3bcf173bdb8R236-R249

**Comment by palmin** [comment]
I updated the date on a commit just to check if CI was in a better mood and now all checks pass.

File: src/userauth.c [hunk 2]
```diff
@@ -118,21 +121,62 @@ static char *userauth_list(LIBSSH2_SESSION *session, const char *username,
...
+                session->userauth_list_state = libssh2_NB_state_idle;
+                return NULL;
+            }
+            if(session->userauth_list_data[0] == SSH_MSG_USERAUTH_BANNER &&
+               (session->userauth_list_data_len >= 5)) {
+
+                methods_len =
+                  _libssh2_ntohu32(session->userauth_list_data + 1);
+
+                /* Cap to 512 bytes. */
```
Comment (willco007) [inline src/userauth.c:146]: The define is 2048, it appears this comment needs updating.

Comment (palmin) [reply src/userauth.c]: Thanks for pointing this out. It should be fixed now.

**Comment by T35R6braPwgDJKq** [comment]
Whats the state here?

**Comment by willco007** [comment]
I was just looking at this again and `userauth_banner` is being leaked. It needs to be freed in `session_free()`

**Comment by willco007** [comment]
It should probably also be set to `NULL` in the if block `session->userauth_list_state == libssh2_NB_state_idle)`. However, that leads me to wonder why this is in the `user_auth` call instead of being its own thing.

**Comment by DevAGCO** [comment]
We are looking forward for this feature to be available again !

**Pull request closed without merging**